packages/buidler-etherscan/src/index.ts

```diff
@@ -1,97 +1,268 @@
+  const bytecodeBuffer = Buffer.from(deployedContractBytecode, "hex");
+  const solcVersionRange = await inferSolcVersion(bytecodeBuffer);
 
-      console.log("Successfully verified contract on etherscan");
+  if (!solcVersionRange.isIncluded(solcVersionConfig)) {
+    let detailedContext;
+    if (solcVersionRange.inferralType === InferralType.EXACT) {
+      detailedContext = `The expected version is ${solcVersionRange}.`;
+    } else {
+      detailedContext = `The expected version range is ${solcVersionRange}.`;
     }
+    const message = `The bytecode retrieved could not have been generated by the selected compiler.
+The selected compiler version is v${config.solc.version}.
+${detailedContext}
+The selected network is ${network.name}.
+Common causes:
+  - Wrong compiler version in buidler config
+  - Wrong address for contract
+  - Wrong network selected or faulty buidler network config`;
+    throw new NomicLabsBuidlerPluginError(pluginName, message);
+  }
+
+  const { lookupMatchingBytecode, compile } = await import("./solc/bytecode");
+  // TODO: this gives us the input for all contracts.
+  // This could be restricted to relevant contracts in a future iteration of the compiler tasks.
+  const { compilerInput, compilerOutput } = await compile(run);
+
+  const contractInformation = await lookupMatchingBytecode(
+    compilerOutput.contracts,
+    deployedContractBytecode,
+    solcVersionRange.inferralType
   );
+  if (contractInformation === null) {
+    const message = `The contract was not found among the ones present in this project.
+The selected network is ${network.name}.
+Common causes:
+  - Wrong address for contract
+  - Wrong network selected or faulty buidler network config`;
+    throw new NomicLabsBuidlerPluginError(pluginName, message);
+  }
+
+  const { Interface } = await import("@ethersproject/abi");
+  const { isABIArgumentLengthError, isABIArgumentTypeError } = await import(
+    "./ABITypes"
+  );
+  const contractInterface = new Interface(contractInformation.contract.abi);
+  let deployArgumentsEncoded;
+  try {
+    deployArgumentsEncoded = contractInterface
+      .encodeDeploy(constructorArguments)
+      .replace("0x", "");
+  } catch (error) {
+    if (isABIArgumentLengthError(error)) {
+      const { contractName, contractFilename } = contractInformation;
+      // TODO: add a list of types and constructor arguments to the error message?
+      const message = `The constructor for ${contractFilename}:${contractName} has ${error.count.types} parameters
+ but ${error.count.values} arguments were provided instead.`;
+      throw new NomicLabsBuidlerPluginError(pluginName, message, error);
+    }
+    if (isABIArgumentTypeError(error)) {
+      const message = `Value ${error.value} cannot be encoded for the parameter ${error.argument}.
+Encoder error reason: ${error.reason}`;
+      throw new NomicLabsBuidlerPluginError(pluginName, message, error);
+    }
+    // Should be unreachable.
```

Comment:
The two guards will probably cover a lot of scenarios, but we don't own ethers so we can't make assumptions about reachability. That being said, if we are reasonably confident that this being reached is an error on our side, we should report it to Sentry. I don't know how likely that is so I don't know if we should do it. (cc @alcuadrado)

Reply:
Can't we check these things with `instanceof`?

Agreed about reporting.